ehrql/measures/measures.py

```diff
@@ -0,0 +1,194 @@
+        }
+        self._validate_kwargs(kwargs)
+
+        # Ensure all arguments are provided (either directly or by defaults)
+        for key, value in kwargs.items():
+            if value is None:
+                raise ValidationError(
+                    f"No value supplied for '{key}' and no default defined"
+                )
+
+        # Ensure measure names are unique
+        if name in self._measures:
+            raise ValidationError(f"Measure already defined with name: {name}")
+
+        # Ensure group_by column definitions are consistent across measures
+        for group_name, definition in get_all_group_by_columns(
+            self._measures.values()
+        ).items():
+            if group_name not in group_by:
+                continue
+            if group_by[group_name].qm_node != definition:
+                raise ValidationError(
+                    f"Inconsistent definition for `group_by` column: {group_name}"
+                )
+
+        # Add measure
+        self._measures[name] = Measure(
+            name=name,
+            numerator=kwargs["numerator"].qm_node,
+            denominator=kwargs["denominator"].qm_node,
+            group_by={
+                group_name: series.qm_node
+                for group_name, series in kwargs["group_by"].items()
+            },
+            intervals=tuple(sorted(set(kwargs["intervals"]))),
+        )
+
+    def _validate_kwargs(self, kwargs):
+        self._validate_num_denom(kwargs["numerator"], "numerator")
+        self._validate_num_denom(kwargs["denominator"], "denominator")
+        self._validate_intervals(kwargs["intervals"])
+        self._validate_group_by(kwargs["group_by"])
+
+    def _validate_num_denom(self, value, name):
+        if value is None:
+            return
+        if not isinstance(value, PatientSeries):
+            raise ValidationError(
+                f"`{name}` must be a one row per patient series,"
+                f" got '{type(value)}': {value!r}"
+            )
+        if value._type not in (bool, int):
+            raise ValidationError(
+                f"`{name}` must be a boolean or integer series, got '{value._type}'"
+            )
+
+    def _validate_intervals(self, intervals):
+        if intervals is None:
+            return
+        if not isinstance(intervals, list):
+            raise ValidationError(
+                f"`intervals` must be a list, got '{type(intervals)}': {intervals!r}"
+            )
+        for interval in intervals:
+            self._validate_interval(interval)
+
+    def _validate_interval(self, interval):
+        if (
+            not isinstance(interval, tuple)
+            or len(interval) != 2
+            or not all(isinstance(i, datetime.date) for i in interval)
+        ):
+            raise ValidationError(
+                f"`intervals` must contain pairs of dates, got: {interval!r}"
+            )
+        if interval[0] > interval[1]:
```

Comment:
I think that's OK . The intervals are closed so that represents just a single date, and if you want to evaluate a measure on a single date then who are we to say you can't?